gensim/matutils.py

```diff
@@ -464,6 +464,82 @@ def cossim(vec1, vec2):
     return result
 
 
+def softcossim(vec1, vec2, similarity_matrix):
+    """Return Soft Cosine Similarity between two vectors given a term similarity matrix.
+
+    Return Soft Cosine Similarity between two sparse vectors given a sparse term similarity matrix
+    in the `scipy.sparse.csc_matrix` format. The similarity is a number between <-1.0, 1.0>, higher
+    is more similar.
+
+    Parameters
+    ----------
+    vec1 : list of (int, float) two-tuples
+        A query vector in the gensim document format.
+    vec2 : list of (int, float) two-tuples of ints
+        A document vector in the gensim document format.
+    similarity_matrix : scipy.sparse.csc_matrix
+        A term similarity matrix.
+
+    Returns
+    -------
+    similarity_matrix.dtype
+        The Soft Cosine Similarity between `vec1` and `vec2`.
+
+    Raises
+    ------
+    ValueError
+        When the term similarity matrix is in an unknown format.
+
+    See Also
+    --------
+    gensim.models.keyedvectors.EuclideanKeyedVectors.similarity_matrix
+        A term similarity matrix produced from term embeddings.
+    gensim.similarities.docsim.SoftCosineSimilarity
+        A class for performing corpus-based similarity queries with Soft Cosine Similarity.
+
+    References
+    ----------
+    Soft Cosine Similarity was perhaps first defined by [sidorovetal14]_.
+
+    .. [sidorovetal14] Grigori Sidorov et al., "Soft Similarity and Soft Cosine Measure: Similarity
+       of Features in Vector Space Model", 2014.
+    """
+
+    def sparse2coo(vec):
+        col = [0] * len(vec)
+        row, data = zip(*vec)
+        return scipy.sparse.coo_matrix((data, (row, col)), shape=(similarity_matrix.shape[0], 1),
+                                       dtype=similarity_matrix.dtype)
+
+    def softdot(vec1, vec2):
+        vec1 = vec1.tocsr()
+        vec2 = vec2.tocsc()
+        return (vec1.T).dot(similarity_matrix).dot(vec2)[0, 0]
+
+    if not isinstance(similarity_matrix, scipy.sparse.csc_matrix):
+        if isinstance(similarity_matrix, scipy.sparse.csr_matrix):
```

Comment:
What's about other formats (sparse, but not `csc` or `csr`)?

Reply:
These would incur a performance hit, because we first need to convert these to `csc` / `csr`. We can either issue a warning and try to convert anything remotely meaningful to the `csc` / `csr` format, which will take a time linear in the number of nonzero entries in `similarity_matrix` and temporarily hold two copies of the (potentially huge) matrix, or we can just refuse anything that is in the wrong format. In the code, I chose the latter approach.